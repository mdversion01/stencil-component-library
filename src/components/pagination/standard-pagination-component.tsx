// src/components/pagination/standard-pagination-component.tsx
import { Component, Prop, Event, EventEmitter, h, Fragment } from '@stencil/core';

@Component({
  tag: 'standard-pagination-component',
  styleUrls: ['./pagination-styles.scss'],
  shadow: false,
})
export class StandardPagination {
  @Prop() currentPage: number = 1;
  @Prop() totalPages: number = 1;
  @Prop() hideGotoEndButtons: boolean = false;
  @Prop() goToButtons: string = '';
  @Prop() size: '' | 'sm' | 'lg' = '';
  @Prop() paginationLayout: '' | 'center' | 'end' | 'fill' | 'fill-left' | 'fill-right' = '';
  @Prop() hideEllipsis: boolean = false;
  @Prop() limit: number = 3;
  @Prop() paginationVariantColor: string = '';
  @Prop() plumage: boolean = false;

  @Event({ eventName: 'change-page' }) changePage!: EventEmitter<{ page: number }>;

  private emitChange(page: number) {
    this.changePage.emit({ page });
  }

  private nextPage = () => {
    if (this.currentPage < this.totalPages) this.emitChange(this.currentPage + 1);
  };
  private prevPage = () => {
    if (this.currentPage > 1) this.emitChange(this.currentPage - 1);
  };
  private firstPage = () => this.emitChange(1);
  private lastPage = () => this.emitChange(this.totalPages);

  private renderEllipsis() {
    if (this.hideEllipsis) return null;
    const liFlex =
      this.paginationLayout === 'fill' ||
      this.paginationLayout === 'fill-left' ||
      this.paginationLayout === 'fill-right'
        ? ' flex-fill d-flex'
        : '';
    const aFlex =
      this.paginationLayout === 'fill' ||
      this.paginationLayout === 'fill-left' ||
      this.paginationLayout === 'fill-right'
        ? ' flex-fill d-flex'
        : '';
    return (
      <li role="separator" class={`page-item disabled bv-d-xs-down-none${liFlex}`}>
        <span class={`page-link${aFlex}`}>...</span>
      </li>
    );
  }

  private renderPageButton(page: number) {
    const liFlex =
      this.paginationLayout === 'fill' ||
      this.paginationLayout === 'fill-left' ||
      this.paginationLayout === 'fill-right'
        ? ' flex-fill d-flex'
        : '';
    const aFlex =
      this.paginationLayout === 'fill' ||
      this.paginationLayout === 'fill-left' ||
      this.paginationLayout === 'fill-right'
        ? ' flex-grow-1'
        : '';
    const active = this.currentPage === page;
    return (
      <li
        role="presentation"
        class={`page-item${active ? ' active' : ''}${liFlex}`}
      >
        <button
          role="menuitemradio"
          type="button"
          aria-controls="pagination"
          aria-label={`Go to page ${page}`}
          aria-checked={active ? ('true' as any) : ('false' as any)}
          tabIndex={active ? -1 : 0}
          class={`page-link${aFlex}`}
          onClick={() => this.emitChange(page)}
          disabled={active}
        >
          {page}
        </button>
      </li>
    );
  }

  private generatePageButtons() {
    const buttons: any[] = [];
    const halfLimit = Math.floor((this.limit - 1) / 2);

    let startPage = Math.max(1, this.currentPage - halfLimit);
    let endPage = Math.min(this.totalPages, startPage + this.limit - 1);

    if (endPage - startPage < this.limit - 1) {
      startPage = Math.max(1, endPage - this.limit + 1);
    }

    if (startPage > 1) {
      buttons.push(this.renderPageButton(1));
      if (startPage > 2) buttons.push(this.renderEllipsis());
    }

    for (let i = startPage; i <= endPage; i++) {
      buttons.push(this.renderPageButton(i));
    }

    if (endPage < this.totalPages) {
      if (endPage < this.totalPages - 1) buttons.push(this.renderEllipsis());
      buttons.push(this.renderPageButton(this.totalPages));
    }

    return buttons;
  }

  render() {
    const sizeCls =
      this.size === 'sm' ? ' pagination-sm' : this.size === 'lg' ? ' pagination-lg' : '';
    const layoutCls =
      this.paginationLayout === 'center'
        ? ' justify-content-center flex-fill50'
        : this.paginationLayout === 'end'
        ? ' justify-content-end'
        : this.paginationLayout === 'fill'
        ? ' text-center'
        : '';
    const colorCls = this.paginationVariantColor ? ` ${this.paginationVariantColor}` : '';
    const plumageCls = this.plumage ? ' plumage' : '';

    const liFlex =
      this.paginationLayout === 'fill' ||
      this.paginationLayout === 'fill-left' ||
      this.paginationLayout === 'fill-right'
        ? ' flex-fill d-flex'
        : '';
    const aFlex =
      this.paginationLayout === 'fill' ||
      this.paginationLayout === 'fill-left' ||
      this.paginationLayout === 'fill-right'
        ? ' flex-grow-1'
        : '';

    return (
      <ul
        role="menubar"
        aria-disabled="false"
        aria-label="Pagination"
        class={`pagination b-pagination${sizeCls}${layoutCls}${colorCls}${plumageCls}`}
      >
        {!this.hideGotoEndButtons ? (
           <Fragment>
            <li
              role="presentation"
              aria-hidden={this.currentPage === 1 ? ('true' as any) : ('false' as any)}
              class={`page-item${this.currentPage === 1 ? ' disabled' : ''}${liFlex}`}
            >
              <button
                role="menuitem"
                type="button"
                tabIndex={this.currentPage === 1 ? -1 : 0}
                aria-label="Go to first page"
                aria-controls="pagination"
                class={`page-link${aFlex}`}
                onClick={this.firstPage}
                disabled={this.currentPage === 1}
              >
                {this.goToButtons === 'text' ? 'First' : '«'}
              </button>
            </li>
            <li
              role="presentation"
              aria-hidden={this.currentPage === 1 ? ('true' as any) : ('false' as any)}
              class={`page-item${this.currentPage === 1 ? ' disabled' : ''}${liFlex}`}
            >
              <button
                role="menuitem"
                type="button"
                tabIndex={this.currentPage === 1 ? -1 : 0}
                aria-label="Go to previous page"
                aria-controls="pagination"
                class={`page-link${aFlex}`}
                onClick={this.prevPage}
                disabled={this.currentPage === 1}
              >
                {this.goToButtons === 'text' ? 'Prev' : '‹'}
              </button>
            </li>
          </Fragment>
        ) : null}

        {this.generatePageButtons()}

        {!this.hideGotoEndButtons ? (
          <Fragment>
            <li
              role="presentation"
              aria-hidden={this.currentPage === this.totalPages ? ('true' as any) : ('false' as any)}
              class={`page-item${this.currentPage === this.totalPages ? ' disabled' : ''}${liFlex}`}
            >
              <button
                role="menuitem"
                type="button"
                tabIndex={this.currentPage === this.totalPages ? -1 : 0}
                aria-label="Go to next page"
                aria-controls="pagination"
                class={`page-link${aFlex}`}
                onClick={this.nextPage}
                disabled={this.currentPage === this.totalPages}
              >
                {this.goToButtons === 'text' ? 'Next' : '›'}
              </button>
            </li>
            <li
              role="presentation"
              aria-hidden={this.currentPage === this.totalPages ? ('true' as any) : ('false' as any)}
              class={`page-item${this.currentPage === this.totalPages ? ' disabled' : ''}${liFlex}`}
            >
              <button
                role="menuitem"
                type="button"
                tabIndex={this.currentPage === this.totalPages ? -1 : 0}
                aria-label="Go to last page"
                aria-controls="pagination"
                class={`page-link${aFlex}`}
                onClick={this.lastPage}
                disabled={this.currentPage === this.totalPages}
              >
                {this.goToButtons === 'text' ? 'Last' : '»'}
              </button>
            </li>
          </Fragment>
        ) : null}
      </ul>
    );
  }
}
