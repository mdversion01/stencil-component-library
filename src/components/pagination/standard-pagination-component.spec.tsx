// src/components/pagination/standard-pagination-component.spec.tsx
import { newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { StandardPagination } from './standard-pagination-component';

describe('standard-pagination-component', () => {
  it('default snapshot with first/prev/next/last', async () => {
    const page = await newSpecPage({
      components: [StandardPagination],
      template: () => <standard-pagination-component currentPage={3} totalPages={9} limit={5} paginationLayout="center" goToButtons="text" />,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('does not render <undefined> wrapper (Fragments handled)', async () => {
    const page = await newSpecPage({
      components: [StandardPagination],
      template: () => <standard-pagination-component currentPage={1} totalPages={3} />,
    });
    expect(page.root!.querySelector('undefined')).toBeNull();
  });

  it('hideGotoEndButtons removes nav edges', async () => {
    const page = await newSpecPage({
      components: [StandardPagination],
      template: () => <standard-pagination-component currentPage={2} totalPages={5} hideGotoEndButtons />,
    });
    expect(page.root).toMatchSnapshot();
    const texts = Array.from(page.root!.querySelectorAll('button.page-link')).map(b => b.textContent?.trim());
    expect(texts).not.toContain('First');
    expect(texts).not.toContain('Prev');
    expect(texts).not.toContain('Next');
    expect(texts).not.toContain('Last');
  });
});
