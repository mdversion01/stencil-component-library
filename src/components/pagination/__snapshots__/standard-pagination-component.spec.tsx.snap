// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`standard-pagination-component default snapshot with first/prev/next/last 1`] = `
<standard-pagination-component>
  <ul aria-disabled="false" aria-label="Pagination" class="b-pagination flex-fill50 justify-content-center pagination" role="menubar">
    <li aria-hidden="false" class="page-item" role="presentation">
      <button aria-controls="pagination" aria-label="Go to first page" class="page-link" role="menuitem" tabindex="0" type="button">
        First
      </button>
    </li>
    <li aria-hidden="false" class="page-item" role="presentation">
      <button aria-controls="pagination" aria-label="Go to previous page" class="page-link" role="menuitem" tabindex="0" type="button">
        Prev
      </button>
    </li>
    <li class="page-item" role="presentation">
      <button aria-checked="false" aria-controls="pagination" aria-label="Go to page 1" class="page-link" role="menuitemradio" tabindex="0" type="button">
        1
      </button>
    </li>
    <li class="page-item" role="presentation">
      <button aria-checked="false" aria-controls="pagination" aria-label="Go to page 2" class="page-link" role="menuitemradio" tabindex="0" type="button">
        2
      </button>
    </li>
    <li class="active page-item" role="presentation">
      <button aria-checked="true" aria-controls="pagination" aria-label="Go to page 3" class="page-link" disabled="" role="menuitemradio" tabindex="-1" type="button">
        3
      </button>
    </li>
    <li class="page-item" role="presentation">
      <button aria-checked="false" aria-controls="pagination" aria-label="Go to page 4" class="page-link" role="menuitemradio" tabindex="0" type="button">
        4
      </button>
    </li>
    <li class="page-item" role="presentation">
      <button aria-checked="false" aria-controls="pagination" aria-label="Go to page 5" class="page-link" role="menuitemradio" tabindex="0" type="button">
        5
      </button>
    </li>
    <li class="bv-d-xs-down-none disabled page-item" role="separator">
      <span class="page-link">
        ...
      </span>
    </li>
    <li class="page-item" role="presentation">
      <button aria-checked="false" aria-controls="pagination" aria-label="Go to page 9" class="page-link" role="menuitemradio" tabindex="0" type="button">
        9
      </button>
    </li>
    <li aria-hidden="false" class="page-item" role="presentation">
      <button aria-controls="pagination" aria-label="Go to next page" class="page-link" role="menuitem" tabindex="0" type="button">
        Next
      </button>
    </li>
    <li aria-hidden="false" class="page-item" role="presentation">
      <button aria-controls="pagination" aria-label="Go to last page" class="page-link" role="menuitem" tabindex="0" type="button">
        Last
      </button>
    </li>
  </ul>
</standard-pagination-component>
`;

exports[`standard-pagination-component hideGotoEndButtons removes nav edges 1`] = `
<standard-pagination-component>
  <ul aria-disabled="false" aria-label="Pagination" class="b-pagination pagination" role="menubar">
    <li class="page-item" role="presentation">
      <button aria-checked="false" aria-controls="pagination" aria-label="Go to page 1" class="page-link" role="menuitemradio" tabindex="0" type="button">
        1
      </button>
    </li>
    <li class="active page-item" role="presentation">
      <button aria-checked="true" aria-controls="pagination" aria-label="Go to page 2" class="page-link" disabled="" role="menuitemradio" tabindex="-1" type="button">
        2
      </button>
    </li>
    <li class="page-item" role="presentation">
      <button aria-checked="false" aria-controls="pagination" aria-label="Go to page 3" class="page-link" role="menuitemradio" tabindex="0" type="button">
        3
      </button>
    </li>
    <li class="bv-d-xs-down-none disabled page-item" role="separator">
      <span class="page-link">
        ...
      </span>
    </li>
    <li class="page-item" role="presentation">
      <button aria-checked="false" aria-controls="pagination" aria-label="Go to page 5" class="page-link" role="menuitemradio" tabindex="0" type="button">
        5
      </button>
    </li>
  </ul>
</standard-pagination-component>
`;
